packages/git/src/browser/git-widget.tsx

```diff
@@ -310,77 +318,89 @@ export class GitWidget extends VirtualWidget implements StatefulWidget {
         return undefined;
     }
 
-    protected renderLastCommit(): h.Child {
+    protected renderLastCommit(): React.ReactNode {
         if (!this.lastCommit) {
             return '';
         }
         const { commit, avatar } = this.lastCommit;
-        const gravatar = h.div({ className: GitWidget.Styles.LAST_COMMIT_MESSAGE_AVATAR }, h.img({ src: avatar }));
-        const summary = h.div({ className: GitWidget.Styles.LAST_COMMIT_MESSAGE_SUMMARY }, commit.summary);
-        const time = h.div({ className: GitWidget.Styles.LAST_COMMIT_MESSAGE_TIME }, `${commit.authorDateRelative} by ${commit.author.name}`);
-        const details = h.div({ className: GitWidget.Styles.LAST_COMMIT_DETAILS }, summary, time);
-        // Yes, a container. Otherwise the button would stretch vertically. And having a bigger `Undo` button than a `Commit` would be odd.
-        const buttonContainer = h.div({ className: GitWidget.Styles.FLEX_CENTER }, h.button({
-            className: `theia-button`,
-            title: 'Undo last commit',
-            onclick: () => this.undo.bind(this)()
-        }, 'Undo'));
-        return VirtualRenderer.flatten([gravatar, details, buttonContainer]);
-    }
-
-    protected renderCommandBar(repository: Repository | undefined): h.Child {
-        const refresh = h.a({
-            className: 'toolbar-button',
-            title: 'Refresh',
-            onclick: async e => {
-                await this.repositoryProvider.refresh();
-            }
-        }, h.i({ className: 'fa fa-refresh' }));
-        const more = repository ? h.a({
-            className: 'toolbar-button',
-            title: 'More...',
-            onclick: event => {
-                const el = (event.target as HTMLElement).parentElement;
-                if (el) {
-                    this.contextMenuRenderer.render(GitWidget.ContextMenu.PATH, {
-                        x: el.getBoundingClientRect().left,
-                        y: el.getBoundingClientRect().top + el.offsetHeight
-                    });
+        return <React.Fragment>
+            <div className={GitWidget.Styles.LAST_COMMIT_MESSAGE_AVATAR}>
+                <img src={avatar} />
+            </div>
+            <div className={GitWidget.Styles.LAST_COMMIT_DETAILS}>
+                <div className={GitWidget.Styles.LAST_COMMIT_MESSAGE_SUMMARY}>{commit.summary}</div>
+                <div className={GitWidget.Styles.LAST_COMMIT_MESSAGE_TIME}>{`${commit.authorDateRelative} by ${commit.author.name}`}</div>
+            </div>
+            <div className={GitWidget.Styles.FLEX_CENTER}>
+                <button className='theia-button' title='Undo last commit' onClick={() => this.undo.bind(this)()}>
+                    Undo
+            </button>
+            </div>
+        </React.Fragment>;
+    }
+
+    protected refreshHandler = async () => {
```

Comment:
A few remarks on the new style:
 - The fields ~can~ should be `readonly`.
 - Could you please move up the fields. I know, I know we do not have a well-defined rule where to put the fields but when just glancing into the code, I would expect to find them above the constructor function. What do you think?
 - It might seem to be an overkill, but could you have methods that return with the handlers? So it would be easier to extend the class. For instance:
```js
protected readonly fooHandler: (sg: Something) => AnotherThing = ...;
protected getFooHandlerOrIfYouHaveABettername(): (sg: Something) => AnotherThing {
  return this.fooHandler;
}

```

Reply:
- these properties are actually functions and should be named as functions, i.e. `refresh`
- since properties cannot be overriden i would leave such properties as not readonly that subclasses can reassign them in the constructor
- for better extensibility where it is need i would call a protected method from such propery

for example:
```ts
// should be used as a react event handler, can be reassigned by subclasses
protected refresh = (event) => this.doRefresh(event);
// should not be used as a react event handler, can be overriden by subclasses
protected doRefresh(event) {
    //...
}
```

Reply:
Yes, this is a way better approach.